community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/LFUCache.scala

```diff
@@ -19,47 +19,31 @@
  */
 package org.neo4j.cypher.internal.compiler.v3_1
 
-import com.googlecode.concurrentlinkedhashmap.ConcurrentLinkedHashMap
+import java.util.function.Function
 
-class LRUCache[K, V](val size: Int) extends ((K, => V) => V) {
+import com.github.benmanes.caffeine.cache.{Cache, Caffeine}
 
-  private val inner = new ConcurrentLinkedHashMap.Builder[K, V]
-    .maximumWeightedCapacity(size)
-    .build()
+class LFUCache[K <: AnyRef, V <: AnyRef](val size: Int) extends ((K, => V) => V) {
 
-  def getOrElseUpdate(key: K, f: => V): V = {
-    val value = inner.get(key)
+  val inner: Cache[K, V] = Caffeine.newBuilder().maximumSize(size).build[K, V]()
 
-    if (value == null) {
-      val createdValue = f
-      val previousValue = inner.putIfAbsent(key, createdValue)
-      if (previousValue == null) createdValue else previousValue
-    } else {
-      value
-    }
-  }
+  def getOrElseUpdate(key: K, f: => V): V = inner.get(key, new Function[K, V] {
+    override def apply(t: K): V = f
+  })
 
-  def getOrElseUpdateByKey(key: K, f: K => V): V = {
-    val value = inner.get(key)
+  def getOrElseUpdateByKey(key: K, f: K => V): V = inner.get(key, new Function[K, V] {
+    override def apply(t: K): V = f(t)
+  })
 
-    if (value == null) {
-      val createdValue = f(key)
-      val previousValue = inner.putIfAbsent(key, createdValue)
-      if (previousValue == null) createdValue else previousValue
-    } else {
-      value
-    }
-  }
-
-  def get(key: K): Option[V] = Option(inner.get(key))
+  def get(key: K): Option[V] = Option(inner.getIfPresent(key))
 
   def put(key: K, value: V) = inner.put(key, value)
 
   def +=(kv: (K, V)) = put(kv._1, kv._2)
 
-  def remove(key: K): Option[V] = Option(inner.remove(key))
+  def remove(key: K): Option[V] = Option(inner.asMap().remove(key))
 
-  def containsKey(key: K) = inner.containsKey(key)
+  def containsKey(key: K) = inner.getIfPresent(key) != null
```

Comment:
You might prefer `inner.asMap().containsKey` to avoid being internally counting as a hit/miss. That effects the eviction policy, but might also skew metrics if you ever enabled stats.